google-cloud-datastore/acceptance/datastore/aggregate_query_test.rb

```diff
@@ -264,9 +293,449 @@
         gql = dataset.gql "SELECT COUNT(*) FROM Character WHERE __key__ HAS ANCESTOR @bookKey",
                           bookKey: book.key
         res = dataset.run_aggregation gql
-        _(res.get).must_equal 3
+        _(res.get).must_be_kind_of Integer
+        _(res.get).must_equal 4
+      end
+    end
+  end
+
+  describe "SUM via API" do
```

Comment:
This is a lot of acceptance ( == integration ) tests for a single feature. Is it possible to verify any of these behaviors with unit tests?

Reply:
These tests focus on the integration between client and server, so they are better placed as acceptance tests, rather than unit tests. They essentially follow the style of "If client sends input `X` with condition`Y`, does the server respond with output `Z`?", with various combinations of `(X, Y, Z)`, including errors.

However, I did a quick scan & identified duplicate tests between all aggregates, and moved them into a common `describe` block.

Reply:
This is fine for now. However, something to consider for the future: integration tests are not cheap. They take time to run, and have increased risk of flakiness. We generally don't need to verify the server's behavior (e.g. "does the server respond with `Z`?") That's the service team's job. Our job is to make sure the client sends the right data, and interprets the server's response correctly. We can usually do that with unit tests, where we mock out the server. Typically we write one or two integration tests for a set of features, just as a "smoke test" to ensure that nothing strange happens when we run things end-to-end, but otherwise try to verify all our feature set with unit tests.